checker-util/src/main/java/org/checkerframework/checker/regex/util/RegexUtil.java

```diff
@@ -334,4 +426,147 @@ private static String regexErrorMessage(String s, int expectedGroups, int actual
+        // when we encountered a '\' which was used to escape something, we traverse till the end
+        // of the escape construct and then only resume normal traversal. This is why we are never
+        // checking if the previous character was a '\'.
+
+        // Optional group here onwards means the ith capturing group, which may not match any part
+        // of a text that matched the regular expression and thus may return null on calls to
+        // matcher.group(i).
+
+        // Special construct means either pure, non-capturing groups that do not capture text and do
+        // not count towards the group total.
+
+        final int length = regexp.length();
+        for (int i = 0; i < length; i++) {
+            // Marks the opening of either a capturing a capturing group or some special
+            // construct.
+            if (regexp.charAt(i) == '(') {
+                boolean isThisCapturingGroup = false;
+                // If '(' is not followed by a '?' or is followed by '?<', it is a capturing group.
+                if ((i < length - 1)
+                        && (regexp.charAt(i + 1) != '?' || regexp.startsWith("?<", i + 1))) {
+                    group += 1;
+                    // Push the group index onto a stack.
+                    unclosedCapturingGroups.push(group);
+                    // This opening bracket marked the opening of a capturing group.
+                    isThisCapturingGroup = true;
+                }
+                // Push the boolean on top of a stack. When we encounter a ')', we will use this
+                // stack to check whether it closes a capturing group or a special construct.
+                isCapturingGroup.push(isThisCapturingGroup);
+                if (isThisCapturingGroup) {
+                    // If this opening was preceded by a '|', this group is optional.
+                    if (i > 0 && regexp.charAt(i - 1) == '|') {
+                        nonNullGroups.remove(Integer.valueOf(group));
+                    }
+                }
+            } else if (regexp.charAt(i) == ')') {
+                // This closes the last opened "thing" (either a capturing group or a special
+                // construct). What the last opened "thing" was, is determined from the boolean
+                // stack.
+                // If the top is false, then it was a special construct.
+                // If the top is true, then it was a capturing group and we need to analyse it
+                // further.
+                boolean closesCapturingGroup = isCapturingGroup.pop();
+                if (closesCapturingGroup) {
+                    // The capturing group being closed is on top of the unclosedCapturingGroups
+                    // stack.
+                    Integer closedGroupIndex = unclosedCapturingGroups.pop();
+                    // If the ')' is followed by a '?', '*', '|' or '{0', then this capturing group
```

Comment:
This is not a correct test.  A simple syntactic test of the nearby characters will not work, because the relevant `'|'` may be arbitrarily far after the `')'`, or may appear before it.  A single `'|'` character may also make multiple groups optional.  I added a couple test cases, but many more devious ones are possible.

Reply:
You might have figured this out already, but one implementation approach would be to add a stack that keeps track of all opened parenthesis-constructs, recording whether that parenthesis-construct contains an alternation `|`.  When a parenthesis-construct is closed, mark any contained capturing groups as optional.  Likewise, when any capturing group is marked as optional because it is followed by `?` etc., mark any contained capturing group as optional.

Reply:
Yes, I completely missed this case. I am really sorry.
I don't see how keeping all opened parenthesis-constructs will help, I think we can do it with the information we already have. The main observation is that when a capturing group is enclosed within another, its group number will be greater than the group number of the enclosing group. So when we close a group with group number `i`, and we find it is immediately followed by one of `? * {0'`, we can mark all the groups with group numbers from `i+1` to `groupCount`. Handling `'|'` will be a little bit different, when we find a group containing `'|'`, everything inside the group will be optional.
Can you please find some bug in this approach and elaborate upon yours and why it might be a better one?
I am sorry for my late follow-ups, my class schedule has been really hectic for the past two weeks.